File: trunk/Geometry/edge_height.cpp
#ifndef EDGEHEIGHT_H_
#define EDGEHEIGHT_H_

#include <cmath>
#include <map>
#include <new>
using namespace std;

#include "miscmath.h"
#include "geoquant.h"
#include "geoquants.h"
#include "triposition.h"


EdgeHeightIndex* EdgeHeight::Index = NULL;

EdgeHeight::EdgeHeight( Edge& e, Face& f ){
  StdFace sf = labelFace( e, f );
  d_ij = PartialEdge::At( sf.v1, sf.e12 );
  d_ik = PartialEdge::At( sf.v1, sf.e13 );
  theta_i = EuclideanAngle::At( sf.v1, f );
  d_ij->addDependent(this);
  d_ik->addDependent(this);
  theta_i->addDependent(this);
}

void EdgeHeight::recalculate(){
  double dij = d_ij->getValue();
  double dik = d_ik->getValue();
  double theta = theta_i->getValue();

  value = dik - dij * cos(theta)/sin(theta);
}

void EdgeHeight::remove() {
     deleteDependents();
     d_ij->removeDependent(this);
     d_ik->removeDependent(this);
     theta_i->removeDependent(this);
     Index->erase(pos);
     delete this;
}

EdgeHeight::~EdgeHeight(){ }

EdgeHeight* EdgeHeight::At( Edge& e, Face& f ){
  TriPosition T( 2, e.getSerialNumber(), f.getSerialNumber() );
  if( Index == NULL ) Index = new EdgeHeightIndex();
  EdgeHeightIndex::iterator iter = Index->find( T );

  if( iter == Index->end() ){
    EdgeHeight* val = new EdgeHeight( e, f );
    val->pos = T;
    Index->insert( make_pair( T, val ) );
    return val;
  } else {
    return iter->second;
  }
}

void EdgeHeight::CleanUp(){
  if( Index == NULL ) return;
  EdgeHeightIndex::iterator iter;
  EdgeHeightIndex copy = *Index;
  for(iter = copy.begin(); iter != copy.end(); iter++) {
    iter->second->remove();
  }
    
  delete Index;
  Index = NULL;
}

#endif /* EDGEHEIGHT_H_ */

File: trunk/Java/src/Geoquant/EdgeHeight.java
package Geoquant;

import java.util.HashMap;

import Triangulation.Edge;
import Triangulation.Face;
import Triangulation.StdFace;

public class EdgeHeight extends Geoquant {
  // Index map
  private static HashMap<TriPosition, EdgeHeight> Index = new HashMap<TriPosition, EdgeHeight>();
  
  // Needed geoquants
  private PartialEdge dij;
  private PartialEdge dik;
  private Angle theta_i;
  
  public EdgeHeight(Edge e, Face f) {
    super(e, f);
    
    StdFace sf = new StdFace(f, e);
    dij = PartialEdge.At(sf.v1, sf.e12);
    dik = PartialEdge.At(sf.v1, sf.e13);
    theta_i = Angle.At(sf.v1, f);
    
    dij.addObserver(this);
    dik.addObserver(this);
    theta_i.addObserver(this);
  }
  
  protected void recalculate() {
    double d_ij = dij.getValue();
    double d_ik = dik.getValue();
    double theta = theta_i.getValue();

    value = (d_ik - d_ij * Math.cos(theta))/Math.sin(theta);
  }
 
  protected void remove() {
    deleteDependents();
    dij.deleteObserver(this);
    dik.deleteObserver(this);
    theta_i.deleteObserver(this);
    Index.remove(pos);
  }
  
  public static EdgeHeight At(Edge e, Face f) {
    TriPosition T = new TriPosition(e.getSerialNumber(), f.getSerialNumber());
    EdgeHeight q = Index.get(T);
    if(q == null) {
      q = new EdgeHeight(e, f);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static double valueAt(Edge e, Face f) {
    return At(e, f).getValue();
  }
  
}
